proposals/0046-optimistic-cluster-restart-automation.md

```diff
@@ -0,0 +1,408 @@
+
+* 5% validators in restart: `PERCENT_NOT_IN_RESTART` is 100% - 5% = 95%.
+`OPTIMISTIC_CONFIRMED_THRESHOLD` - `MALICIOUS_SET` - `PERCENT_NOT_IN_RESTART`
+= 67% - 5% - 95% < 10%, so no validators would repair any block.
+
+* 70% validators in restart: `PERCENT_NOT_IN_RESTART` is 100% - 70% = 30%.
+`OPTIMISTIC_CONFIRMED_THRESHOLD` - `MALICIOUS_SET` - `PERCENT_NOT_IN_RESTART`
+= 67% - 5% - 30% = 32%, so slots with above 32% votes accumulated from
+`RestartLastVotedForkSlots` would be repaired.
+
+* 80% validators are in restart, `PERCENT_NOT_IN_RESTART` is 100% - 80% = 20%.
+`OPTIMISTIC_CONFIRMED_THRESHOLD` - `MALICIOUS_SET` - `PERCENT_NOT_IN_RESTART`
+= 67% - 5% - 20% = 42%, so slots with above 42% votes accumulated from
+`RestartLastVotedForkSlots` would be repaired.
+
+From above examples, we can see the "must-have" threshold changes dynamically 
+depending on how many validators are in restart. The main benefit is that a
+block will only move from "must-have" to "ignored" as more validators 
+join the restart, not vice versa. So the list of blocks a validator needs to
+repair will never grow bigger when more validators join the restart.
+
+Once the validator gets `RestartLastVotedForkSlots`, it can calculate which
+blocks must be repaired. When all those "must-have" blocks are repaired and
+replayed, it can proceed to step 3.
+
+3. **gossip current heaviest fork**
+
+The main goal of this step is to "vote" the heaviest fork to restart from.
+
+We use a new gossip message `RestartHeaviestFork`, its fields are:
+
+* `slot`: `u64` slot of the picked block.
+* `hash`: `Hash` bank hash of the picked block.
+* `stake_committed_percent`: `u16` total percentage of stakes of the validators
+it received `RestartHeaviestFork` messages from.
+
+After receiving `RestartLastVotedForkSlots` from the validators holding stake
+more than `RESTART_STAKE_THRESHOLD` and repairing slots in "must-have"
+category, replay all blocks and pick the heaviest fork by traversing from
+local root like this:
+
+1. If a block has more than 67% stake in `RestartLastVotedForkSlots`
+   messages, traverse down this block.
+
+2. Define the "must have" threshold to be 62%. If traversing to a block with
+   more than one child, we check for each child
+   `vote_on_child + stake_on_validators_not_in_restart >= 62%`. If so, traverse
+   to the child.
+
+   For example, if 80% validators are in restart, child has 42% votes, then
+   42 + (100-80) = 62%, traverse to this child. 62% is chosen instead of 67%
+   because 5% could make the wrong votes.
+
+   Otherwise stop traversing the tree and use last visited block.
+
+To see why the above algorithm is safe, assuming one block X is optimistically
+confirmed before the restart, we prove that it will always be either the block
+picked or the ancestor of the block picked.
+
+Assume X is not picked, then:
+
+1. If its parent Y is on the Heaviest fork, then either a sibling of X is
+   chosen or no child of Y is chosen. In either case
```

Comment:
"no child of `Y` is chosen" I think can be simplified to "Y is chosen"

Reply:
Y is chosen is assumption. Here are two cases when X is not chosen:
1. no child is picked, Y is the last one on the picked fork
2. a sibling of X is chosen.
No child of Y is chosen corresponds to case 1.